### ISSUE TITLE
Enhancement: [array-type] Allow Promise<T>[]

### ISSUE THREAD

**Issue by JasonMan34** [typescript-eslint/typescript-eslint#10968]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/array-type/

### Description

Using the `{ default: 'array-simple' }` option, this pattern currently gets flagged by the rule:

```ts
const promises: Promise<MyType>[] = [];
// Populate the array
```

This is the async counterpart of

```ts
const values: MyType[] = [];
// Populate the array
```

Could an option to not flag promises this way potentially be added?

### Fail

```TypeScript
const promises: Array<Promise<MyType>> = [];
```

### Pass

```TypeScript
const promises: Promise<MyType>[] = [];
```

**Comment by kirkwaiblinger** [comment]
Hi @JasonMan34! I'm wondering whether a review of the docs is in order or there is some other misunderstanding...

> This pattern currently gets flagged by the rule:
> 
> const promises: Promise<MyType>[] = [];
> // Populate the array

With what option? Not with the default option, which always enforces the `T[]` syntax. ([playground](https://typescript-eslint.io/play/#ts=5.8.2&fileType=.ts&code=C4TwDgpgBAsiAq5oF4oEMB2IDcAoAxgPYYDOwUYAToQLYCWJEJAXFAArX2MA8cikAPgDaAXSipR2IA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AQ2mncNqNPRgA2uGw4A7pyaQANKLFYx2SABNEAM3YJ8gyJ26FI87AF9jAXVFmTQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))

> This is the async counterpart of
> 
> const values: MyType[] = [];
> // Populate the array
> This also applies to async functions that return `Promise<MyType>[]`.

I don't follow this at all. [An `async` function can only ever return a Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function). Therefore, TS correctly requires you to provide a return type of `Promise<T[]>`, never a `Promise<T>[]` ([playground](https://typescript-eslint.io/play/#ts=5.8.2&fileType=.ts&code=IYZwngdgxgBAZgV2gFwJYHsL3egFAShgG8AoAejJiupuoD0B%2BEqgJwFNkEWsBtARgA0MAExCAzAF0SAXxIlQkWIhQYsEdMgCCAG23oA7mwAmBAFwwACi3QBbVCDYAeCAhsAjNiwB8PCcWYw7JzcMPxCojCSMkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AQ2mncNqNPRgA2uGw4A7pyaQANKLFYx2SABNEAM3YJ8gyJ26FI87AF9jAXVFmTQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))

*** 

If, after reviewing the docs and rule options, you feel there's still something here, could you please provide more specific examples with working playground demos? Thanks!

**Comment by JasonMan34** [comment]
@kirkwaiblinger Thank you for the quick response, and apologies for the mistakes in my initial post. I have edited out the parts about async functions and clarification that I'm talking about the `{ default: 'array-simple' }` option

**Comment by JasonMan34** [comment]
> Why do you want to special case promises? What's the difference between `Foo<T>[]` and `Promise<T>[]`?
> 
> IMO they shouldn't be special cased as they are equivalent.

I was only thinking about promises as they are the main culprit that I find this flag with. I'm also not sure if the behavior that I always want for promise wrappers to behave the same in generics, e.g.:

```ts
// Which is preferred?
type Something1 = WithId<Partial<SomeType>>>[]
type Something2 = Array<WithId<Partial<SomeType>>>>
```

However I can see the case for all generics and not just promises, as it's kind of weird to single it out

**Comment by bradzacher** [comment]
That's really the point of `array-simple` - for simple types (namely just bare identifiers) use the array syntax. For all other types use the array generic. Adding exceptions like "except if the generic name is `Promise`" makes the behaviour pretty weird.

---

The rule has existed with this setup for quite some time. So I don't believe this request satisfies the "useful to the broader TypeScript community" check.

And additionally without a solid justification as to why it should be special cased - I don't think the request meets the requirements to be accepted.

For these reasons I am rejecting this proposal.
